src/source/canvas_source.js

```diff
@@ -101,6 +106,21 @@ class CanvasSource extends ImageSource {
      */
     // setCoordinates inherited from ImageSource
 
+    readCanvas() {
+        // We *should* be able to use a pure HTMLCanvasElement in
+        // texImage2D/texSubImage2D (in ImageSource#_prepareImage), but for
+        // some reason this breaks the map on certain GPUs (see #4262).
+
+        if (this.context instanceof CanvasRenderingContext2D) {
+            return this.context.getImageData(0, 0, this.width, this.height);
+        } else if (this.context instanceof WebGLRenderingContext) {
+            const gl = this.context;
+            const data = new Uint8Array(this.width * this.height * 4);
+            gl.readPixels(0, 0, this.width, this.height, gl.RGBA, gl.UNSIGNED_BYTE, data);
+            return new window.ImageData(new Uint8ClampedArray(data), this.width, this.height);
```

Comment:
[According to MDN](https://developer.mozilla.org/en-US/docs/Web/API/ImageData), the `ImageData` constructor is not available on IE (and Safari support is unknown).

One possible way around this would be to return `{width: number, height: number, data: Uint8ClampedArray}` (which `ImageData` will also satisfy), and have `_prepareImage` accept that.

Reply:
I started going down this path but ran into complications with different `texImage2D`/`texSubImage2D` signatures and with flow 😕 instead I went with using `CanvasRenderingContext2D.createImageData`, which is much cleaner but requires creating an extra canvas context for webgl-based canvas sources…